src/query/changed.go

```diff
@@ -0,0 +1,94 @@
+package query
+
+import (
+	"core"
+	"scm"
+)
+
+// ChangedRequest is a simple parameter object
+type ChangedRequest struct {
+	Since            string
+	DiffSpec         string
+	IncludeDependees string
+}
+
+// ChangedLabels returns all BuildLabels that have changed according to the given request.
+func ChangedLabels(state *core.BuildState, request ChangedRequest) []core.BuildLabel {
+	workspaceChangedFiles := changedFiles(request.Since, request.DiffSpec)
+	if len(workspaceChangedFiles) == 0 {
+		return []core.BuildLabel{}
+	}
+
+	labels := make([]core.BuildLabel, 0)
+	targets := targetsForChangedFiles(state.Graph, workspaceChangedFiles, request.IncludeDependees)
+	for _, t := range targets {
+		if state.ShouldInclude(t) {
+			labels = append(labels, t.Label)
+		}
+	}
+
+	return labels
+}
+
+func changedFiles(since string, diffSpec string) []string {
+	if diffSpec != "" {
+		return scm.ChangesIn(diffSpec, "")
+	}
+
+	if since == "" {
+		since = scm.CurrentRevIdentifier()
+	}
+
+	return scm.ChangedFiles(since, true, "")
+}
+
+func targetsForChangedFiles(graph *core.BuildGraph, files []string, includeDependees string) []*core.BuildTarget {
+	addresses := make([]*core.BuildTarget, 0)
+	for _, target := range graph.AllTargets() {
+		for _, source := range target.Sources {
+			if source.Label() == nil {
```

Comment:
you might find `target.HasSource` or `target.HasAbsoluteSource` useful - it should handle the source loops here for you.